### PR TITLE
187718437 v3 DI Create Component

### PR DESCRIPTION
**Pull request by tealefristoe**
PT Story: https://www.pivotaltracker.com/story/show/187718437

This is a massive PR that got a little out of control. Sorry it's so large and unwieldy.

At its core, this PR implements a handler for `create component` API requests for the following tile types:
`calculator`, `caseCard`, `game`, `graph`, `map`, `slider`, `webView`
However, a lot of extra work was needed to get several of these to work, so code in many parts of the codebase ended up being touched.

**General Changes**
- I added `cannotClose` to the tile model and only show the close button on the tile if `cannotClose` is not true. This doesn't work quite like in v2 (in v2 the space for the close button remains, but here the minimize button is shifted over), but I figured this was close enough for this PR.
- New tile options have been expanded, and several functions that always or sometimes create new tiles now accept these options.
- Tile creation functionality has been consolodated into `createTileSnapshotOfType` and `createTileOfType`. These functions are in a new file, `create-tile.ts`, to avoid import cycles.

**Calculator**
This one was straight forward and didn't require any special work.

**caseCard**
The `caseCard` and `caseTable` have a close relationship, and I had to make this more explicit to get `create component caseCard` to work.
- Several functions that were previously just for the `caseTable` now work for the `caseCard` as well, and have been renamed to show this.
- `case-table-utils.ts` include a couple functions that have been expanded to handle both the `caseCard` and `caseTable`, so should be moved into a neutral space from the `case-table` directory. I haven't done that in this PR so it's more clear what I actually changed, but I left comments indicating that these should move in the future.
- Similarly, `DocumentContent.toggleCardTable` should be moved into a utility file rather than being in the generic document content, but I haven't done that in this PR to make it easier to see what I actually changed in the function.
- I made it so `createOrShowTableOrCardForDataset` (formerly without the `OrCard`) always shows the specified existing tile, rather than toggling it like it did previously.

**Graph**
- `GraphController.handleAttributeAssignment` has been removed, and its functionality has been moved into `GraphController.updateGraph` (formerly `initializeGraph`). `updateGraph` is called whenever the `attributeDescriptions` change, including whenever `handleAttributeAssignment` was explicitly called previously. It was necessary to move at least some of the functionality in order to create a graph using a snapshot. However, this is a big change that could have consequences I don't anticipate, so it's worth taking a careful look at the change.
- `GraphDataConfigurationModel.setPrimaryRole` has been changed to `withoutUndo` because it's only called automatically through `GraphController.updateGraph`. When it was added to the undo stack, it was destroying the redo stack when undoing a change of attributes, which was discovered via cypress tests.

**Slider**
- A departure from v2, it's no longer possible to create multiple sliders for the same global value.
- If no global value is specified when creating a slider, a new global value will be created for it (just like if the user used the slider button to create a slider).

**WebView/Game**
These are both treated in the same way and are both straight forward and simple.

**Unimplemented Features**
I wasn't able to implement some features in this PR.
- `text`, `guide`, and `imageView` tiles cannot be created yet.
- The general `position` option is not supported yet.
- `horizontalScrollOffset` and `isIndexHidden` for `create component table`.

### CODE REVIEW

**Comment by kswenson** [comment]
@bfinzer I'm adding you as a reviewer, specifically to take a look at the proposed changes to the graph, graph controller, etc.

**Review by kswenson — CHANGES_REQUESTED** [review]
👍 Generally looks good -- a smattering of small suggestions, but my main concern is with the graph changes, which seem likely to have unforeseen and perhaps undesirable consequences. I'm interested to hear @bfinzer's review, but in the meantime, here's some thoughts I posted on slack as well as a suggestion for how we might move forward:

>A few thoughts on the graph changes before I break for lunch. Prior to the PR, `initializeGraph()` (perhaps not the best name) is called in response to any change to attribute assignment to update the axis scales and other non-serialized properties, i.e. it never actually changes the model. `handleAttributeAssignment()` is only called in response to direct user action (dropping an attribute or selecting `Treat as …`) and its job is explicitly to update the model to make it consistent with the new attribute configuration by updating things like primary role, plot type, and axis types/models. It seems that the impetus for these changes is that the plugin API requires the kind of model synchronization that is currently performed by `handleAttributeAssignment()`, but broader because it can’t assume that only one attribute assignment has changed. But my initial inclination is to think that we should keep the two functions separate, but perhaps rename them to be clearer — perhaps `syncAxisScalesWithModel()` (for `initializeGraph()`) and `syncModelWithAttributeConfiguration()` (for `handleAttributeAssignment()`).
>There are times that we don’t want to make model changes, e.g. restoring documents, undo/redo, etc., where `syncAxisScalesWithModel()` seems more appropriate.

Some thoughts about a possible way forward:
- Move the model-updating code from the new `updateGraph()` method (most of which originated in the former `handleAttributeAssignment()` method) into a utility function `syncModelWithAttributeConfiguration()` (in `graph-model-utils.ts`?) which takes a graph model as argument and updates it appropriately.
- Call this utility function from the existing `handleAttributeAssignment()` method of the graph controller.
- In the DI code, create a local `GraphModel`, then immediately pass it to `syncModelWithAttributeConfiguration()` to configure it appropriately.
- Pass `getSnapshot()` of the local graph model to the create component code.

In other words, use the utility function to get an appropriately configured snapshot so that no further model changes are required after model creation.

File: v3/src/components/graph/models/graph-data-configuration-model.ts [hunk 1]
```diff
@@ -579,6 +580,7 @@ export const GraphDataConfigurationModel = DataConfigurationModel
   })
   .actions(self => ({
     setPrimaryRole(role: GraphAttrRole) {
+      withoutUndo()
```
[the remark this answers is not in the document]
Comment (kswenson) [reply v3/src/components/graph/models/graph-data-configuration-model.ts:583]: Also note that some of the refactoring suggested in review may obviate the need for this.

**Review by bfinzer — APPROVED** [review]
👍🏻LGTM
I spent quite awhile trying to break it and wasn't able to do so. I think Kirk's suggestions for rename and refactoring are important to carry out.

File: v3/src/models/document/document-content.ts [hunk 2]
```diff
@@ -219,27 +208,33 @@ export const DocumentContentModel = BaseDocumentContentModel
     }
   }))
   .actions(self => ({
+    // TODO This action involves specific tile types, so it should not be in the generic document content model.
+    // I'm leaving it here for now so it's clear how it's changed in this PR.
```
[the remark this answers is not in the document]
Comment (tealefristoe) [reply v3/src/models/document/document-content.ts:212]: I think this action should be a function in `case-table-card-utils.ts`. Does that seem reasonable?

File: v3/src/components/case-table/case-table-utils.ts [hunk 3]
```diff
@@ -55,24 +72,31 @@ export const createTableForDataset = (model: ISharedDataSet, caseMetadata: IShar
...
   const { content } = document
   const manager = getSharedModelManager(document)
   const caseMetadatas = manager?.getSharedModelsByType<typeof SharedCaseMetadata>(kSharedCaseMetadataType)
 
-  const model = manager?.getSharedModelsByType("SharedDataSet")
-    .find(m => m.id === sharedDataset.id) as ISharedDataSet | undefined
-  const caseMetadata = caseMetadatas?.find(cm => cm.data?.id === model?.dataSet.id)
-  if (!model || !caseMetadata) return
+  const sharedDataSet = manager?.getSharedModelsByType("SharedDataSet")
+    .find(m => m.id === _sharedDataSet.id) as ISharedDataSet | undefined
```
[the remark this answers is not in the document]
Comment (tealefristoe) [reply v3/src/components/case-table/case-table-utils.ts:84]: I'm not sure why the `sharedDataSet` is searched for, even though it's also a parameter. All the jest tests passed after simplifying it, so let's give it a try.

File: v3/src/data-interactive/data-interactive-component-types.ts [hunk 4]
```diff
@@ -30,19 +31,88 @@ export const kComponentTypeV3ToV2Map: Record<string, string> = {
...
+  type: "text"
+}
+export interface V2WebView extends V2Component {
+  type: "webView"
+  URL?: string
 }
 
-export type V2Component = V2CaseTable
+export type V2SpecificComponent = V2Calculator | V2CaseCard | V2CaseTable | V2Game | V2Graph | V2Guide | V2Map |
+  V2Slider | V2Text | V2WebView
```
[the remark this answers is not in the document]
Comment (tealefristoe) [reply v3/src/data-interactive/data-interactive-component-types.ts:118]: I don't think there's much overlap between these types and those defined in `codap-v2-types.ts`. Even the `type` fields are different.

File: v3/src/data-interactive/handlers/component-handler-slider.test.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    // Delete slider tile
+    const deleteResult = handler.delete!({ component: newValueTile })
+    expect(deleteResult.success).toBe(true)
+    expect(documentContent.tileMap.size).toBe(0)
+
+    // Create slider with value and other options specified
+    const oldValueResult = handler.create!({}, {
+      type: "slider", globalValueName: "global", lowerBound: -100, upperBound: 100,
+      animationDirection: 0, animationMode: 1
+    })
```
[the remark this answers is not in the document]
Comment (tealefristoe) [reply v3/src/data-interactive/handlers/component-handler-slider.test.ts:37]: Yes, v2 allows you to specify the global value, which I believe is the only way to create a slider for a global created via the `create global` API call. It's definitely a bit of a funky system, especially since closing the slider deletes the global in v2, but that's how it works, so that's what I replicated.

**Review by bfinzer — APPROVED** [review]
👍🏻LGTM
Nothing jumped out at me as problematic.
As per our discussion in standup, please go ahead and merge this PR.

**Review by kswenson — COMMENTED** [review]
👍 Changes look good -- nice work with the provisional references.
